src/java/picard/util/JavascriptFilter.java

```diff
@@ -0,0 +1,215 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package picard.util;
+
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.List;
+
+import javax.script.Bindings;
+import javax.script.Compilable;
+import javax.script.CompiledScript;
+import javax.script.ScriptEngine;
+import javax.script.ScriptEngineManager;
+import javax.script.ScriptException;
+import javax.script.SimpleBindings;
+
+import htsjdk.samtools.SAMFileHeader;
+import htsjdk.samtools.SAMRecord;
+import htsjdk.samtools.filter.SamRecordFilter;
+import htsjdk.samtools.util.CloserUtil;
+import htsjdk.samtools.util.CollectionUtil;
+import htsjdk.samtools.util.IOUtil;
+import htsjdk.variant.variantcontext.VariantContext;
+import htsjdk.variant.vcf.VCFFilterHeaderLine;
+import htsjdk.variant.vcf.VCFHeader;
+import picard.vcf.filter.VariantFilter;
+
+/**
+ * Javascript filter with HEADER type containing TYPE records. contains two
+ * static method to get a SAM Read filter or a VariantFilter
+ *
+ */
+public abstract class JavascriptFilter<HEADER, TYPE> {
+	public static final String HEADER_KEY = "header";
+	/** compiled user script */
+	private CompiledScript script = null;
+
+	/** javascript engine */
+	private ScriptEngine engine = null;
+
+	/** javascript bindings */
+	protected Bindings bindings;
+
+	/** put header in the bindings */
+	protected JavascriptFilter(final HEADER h) {
+		this.bindings = new SimpleBindings();
+		this.bindings.put(HEADER_KEY, h);
+	}
+
+	/** returns key used for header binding */
+	public String getHeaderKey() {
+		return HEADER_KEY;
+	}
+
+	/** returns key used for record binding */
+	public abstract String getRecordKey();
+
+	/** Evaluates this predicate on the given argument. */
+	protected boolean accept(final TYPE record) {
+		try {
+			this.bindings.put(getRecordKey(), record);
+			Object result = this.script.eval(bindings);
+			if (result == null)
+				return false;
+			;
+			if (result instanceof Boolean) {
+				if (Boolean.FALSE.equals(result))
+					return false;
+			} else if (result instanceof Number) {
+				if (((Number) result).intValue() != 1)
+					return false;
+			} else {
+				return false;
+			}
+			return true;
+		} catch (ScriptException err) {
+			throw new RuntimeException(err);
+		}
+	}
+
+	/** creates the Javascript manager, read and compile the scriptFile */
+	protected void initialize(final File scriptFile) {
+		FileReader scriptReader = null;
+		final ScriptEngineManager manager = new ScriptEngineManager();
+		this.engine = manager.getEngineByName("js");
+		if (this.engine == null) {
+			throw new RuntimeException("The embedded 'javascript' engine is not available in java. "
```

Comment:
I don't know very much about java javascript engines.  From what I can tell most jvms should include some js engine, either rhino or nashorn.  Things like android might be an exception, but I think anyone running picard on android can be expected to put a bit of work in for compatibility.    I think we should make this exception more specific so someone could attempt to catch it should they want to have different error reporting.  Could you add a new subclass of `PicardException` called `JavaScriptException` and throw that instead.